File: backend/app/services/evidence_builder.py
"""Evidence pack builder — scores, selects, excerpts, and classifies evidence."""

import re
import logging
from dataclasses import dataclass, field

from app.db.models import Artifact

logger = logging.getLogger(__name__)

RATIONALE_KEYWORDS = {
    "because", "since", "due to", "reason", "rationale", "workaround",
    "legacy", "backward", "compatibility", "deprecat", "migrate",
    "issue", "bug", "fix", "patch", "hotfix", "regression", "edge case",
    "tradeoff", "trade-off", "decision", "adr", "consensus",
}


@dataclass
class EvidenceEntry:
    artifact_id: str
    artifact_type: str
    artifact_title: str
    artifact_url: str
    artifact_date: str
    relevance_score: float
    excerpt: str
    claim: str
    is_direct: bool


@dataclass
class EvidencePack:
    entries: list[EvidenceEntry] = field(default_factory=list)
    coverage: str = "insufficient"
    gaps: list[str] = field(default_factory=list)
    total_sources: int = 0


class EvidencePackBuilder:
    def build(self, artifacts: list[Artifact], scores: dict, sources: dict, query_text: str) -> EvidencePack:
        pack = EvidencePack()
        if not artifacts:
            pack.gaps.append("No artifacts found matching the query")
            return pack

        candidates = []
        for artifact in artifacts:
            aid = str(artifact.id)
            base_score = scores.get(aid, 0.5)
            combined = self._compute_score(artifact, base_score, query_text)
            excerpt = self._extract_excerpt(artifact, query_text)
            claim = self._formulate_claim(artifact, query_text)
            candidates.append(EvidenceEntry(
                artifact_id=aid, artifact_type=artifact.artifact_type,
                artifact_title=artifact.title or "Untitled",
                artifact_url=artifact.url or "",
                artifact_date=str(artifact.date) if artifact.date else "",
                relevance_score=combined, excerpt=excerpt, claim=claim,
                is_direct=sources.get(aid, "") == "exact_match",
            ))

        candidates.sort(key=lambda x: x.relevance_score, reverse=True)
        selected = candidates[:12]
        if len(selected) > 3 and selected[-1].relevance_score < 0.2:
            selected = [e for e in selected if e.relevance_score >= 0.2][:max(3, len(candidates))]

        pack.entries = selected
        pack.total_sources = len(candidates)
        direct_count = sum(1 for e in selected if e.is_direct)
        high_relevance = sum(1 for e in selected if e.relevance_score >= 0.5)

        if direct_count >= 1 and high_relevance >= 2:
            pack.coverage = "sufficient"
        elif len(selected) >= 2:
            pack.coverage = "partial"
        else:
            pack.coverage = "insufficient"

        types = {e.artifact_type for e in selected}
        if "pr" not in types and "issue" not in types:
            pack.gaps.append("No pull requests or issues found")
        if "commit" not in types:
            pack.gaps.append("No commit history found")
        return pack

    def _compute_score(self, artifact: Artifact, base: float, query: str) -> float:
        score = base
        if artifact.artifact_type in ("pr", "issue"):
            score += 0.15
        if artifact.artifact_type == "adr":
            score += 0.2
        text = f"{artifact.title or ''} {artifact.content or ''}".lower()
        for kw in RATIONALE_KEYWORDS:
            if kw in text:
                score += 0.05
        if artifact.title:
            terms = [t for t in query.split() if len(t) > 3]
            if any(t.lower() in artifact.title.lower() for t in terms):
                score += 0.1
        if artifact.metadata_ and artifact.metadata_.get("merged"):
            score += 0.1
        return min(score, 1.0)

    def _extract_excerpt(self, artifact: Artifact, query: str, max_len: int = 300) -> str:
        content = artifact.content or artifact.description or artifact.title or ""
        if not content:
            return ""
        paras = [p.strip() for p in re.split(r"\n\s*\n", content) if p.strip()] or [content]
        query_terms = set(query.lower().split())
        best = max(paras, key=lambda p: sum(1 for t in query_terms if t in p.lower()))
        return best[:max_len] + "..." if len(best) > max_len else best

    def _formulate_claim(self, artifact: Artifact, query: str) -> str:
        claims = {
            "commit": f"Commit '{artifact.title}' shows a code change",
            "pr": f"Pull request '{artifact.title}' represents a proposed change",
            "issue": f"Issue '{artifact.title}' discusses a topic",
            "adr": f"Architecture decision record '{artifact.title}' documents a decision",
            "doc": f"Documentation '{artifact.title}' provides context",
            "release_note": f"Release '{artifact.title}' announces changes",
        }
        return claims.get(artifact.artifact_type, f"Artifact '{artifact.title}' provides information")

File: src/lib/services/evidence-builder.ts
import * as db from '@/lib/db';
import * as types from '@/types';
import { getRouter } from '@/lib/providers/router';

/**
 * EvidencePackBuilder constructs compact, high-relevance evidence packs
 * from retrieved artifacts. It removes redundancy, preserves source metadata,
 * and classifies coverage.
 */
export class EvidencePackBuilder {
  /**
   * Build an evidence pack from retrieved artifacts and the original query.
   * Uses LLM to extract relevant excerpts and classify coverage.
   */
  async build(
    artifacts: Array<types.Artifact & { relevanceScore: number; matchType: string }>,
    queryText: string,
    answerId?: string
  ): Promise<types.EvidencePack> {
    const entries: types.EvidencePack['entries'] = [];
    const router = getRouter();

    // Score and select best artifacts
    const scored = artifacts
      .map(a => ({
        artifact: a,
        score: this.computeCombinedScore(a, queryText),
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, 12); // Process top 12, keep 3-8

    for (const { artifact, score } of scored) {
      const excerpt = this.extractRelevantExcerpt(artifact, queryText);
      const claim = this.formatClaim(artifact, excerpt);

      entries.push({
        artifact: artifact as unknown as types.Artifact,
        relevanceScore: score,
        excerpt,
        claim,
        isDirect: score > 0.7,
      });

      // Store in DB if answer exists
      if (answerId) {
        try {
          await db.insertEvidence({
            answer_id: answerId,
            artifact_id: artifact.id,
            relevance_score: score,
            excerpt,
            claim,
            citation_url: artifact.url || undefined,
            is_direct: score > 0.7,
            metadata: {
              artifact_type: artifact.artifact_type,
              author: artifact.author,
              date: artifact.date,
              match_type: artifact.matchType,
            },
          });
        } catch { /* silently continue */ }
      }
    }

    // Determine coverage
    const coverage = this.classifyCoverage(entries);
    const gaps = this.identifyGaps(entries, queryText);

    return {
      entries: entries.slice(0, 8),
      coverage,
      gaps,
      totalSources: artifacts.length,
    };
  }

  private computeCombinedScore(
    artifact: types.Artifact & { relevanceScore: number; matchType: string },
    queryText: string
  ): number {
    let score = artifact.relevanceScore;

    // Boost PRs and issues (they contain rationale)
    if (artifact.artifact_type === 'pr' || artifact.artifact_type === 'issue') {
      score += 0.15;
    }

    // Boost ADRs (they explicitly contain decisions)
    if (artifact.artifact_type === 'adr') {
      score += 0.2;
    }

    // Boost artifacts with explicit rationale keywords
    const content = ((artifact.title || '') + ' ' + (artifact.description || '') + ' ' + (artifact.content || '')).toLowerCase();
    const rationaleWords = ['because', 'reason', 'tradeoff', 'constraint', 'issue', 'fix', 'bug', 'deprecat', 'migrat', 'breaking'];
    for (const word of rationaleWords) {
      if (content.includes(word)) {
        score += 0.05;
      }
    }

    // Boost if query terms appear in title
    const queryTerms = queryText.toLowerCase().split(/\s+/);
    for (const term of queryTerms) {
      if (term.length > 3 && (artifact.title || '').toLowerCase().includes(term)) {
        score += 0.1;
      }
    }

    return Math.min(1.0, score);
  }

  private extractRelevantExcerpt(artifact: types.Artifact, queryText: string): string {
    let content = artifact.content || artifact.description || '';
    content = this.cleanContentForDisplay(content, artifact);

    const queryTerms = queryText.toLowerCase().split(/\s+/);

    // Find the most relevant paragraph
    const paragraphs = content.split('\n\n');
    let bestParagraph = paragraphs[0] || '';
    let bestScore = 0;

    for (const para of paragraphs) {
      const lower = para.toLowerCase();
      let score = 0;
      for (const term of queryTerms) {
        if (term.length > 2 && lower.includes(term)) score++;
      }
      if (score > bestScore) {
        bestScore = score;
        bestParagraph = para;
      }
    }

    return bestParagraph.substring(0, 500).trim();
  }

  /**
   * Clean content from JSON/binary format to human-readable text.
   */
  private cleanContentForDisplay(content: string, artifact: types.Artifact): string {
    if (!content) return '';

    // Try to parse as JSON and extract readable parts
    try {
      const parsed = JSON.parse(content);

      // Commit message JSON
      if (parsed.message && typeof parsed.message === 'string') {
        let text = parsed.message;
        if (Array.isArray(parsed.files) && parsed.files.length > 0) {
          text += '\n\nChanged files:\n' + parsed.files.map((f: any) =>
            `  ${f.path || f.filename || 'unknown'} (${f.status || 'modified'})`
          ).join('\n');
        }
        return text;
      }

      // Query/intent JSON from source files
      if (parsed.query) {
        return `Query: ${parsed.query}`;
      }

      // Generic: extract any string values
      const strings: string[] = [];
      for (const [key, val] of Object.entries(parsed)) {
        if (typeof val === 'string' && val.length > 5 && val.length < 500) {
          strings.push(`${key}: ${val}`);
        }
      }
      if (strings.length > 0) return strings.join('\n');
    } catch { /* not JSON, continue */ }

    // Remove code fences
    let cleaned = content.replace(/```json\s*/g, '').replace(/```\s*/g, '').trim();

    // Remove lines that are just JSON syntax
    cleaned = cleaned.replace(/^[{}\[\]",:]+\s*$/gm, '');

    // Remove leading/trailing JSON brackets
    cleaned = cleaned.replace(/^[{\[]\s*/, '').replace(/\s*[}\]]$/, '');

    return cleaned || content;
  }

  private formatClaim(artifact: types.Artifact, excerpt: string): string {
    const typeLabel = artifact.artifact_type.toUpperCase();
    const title = artifact.title || 'Untitled';
    const author = artifact.author ? ` by ${artifact.author}` : '';
    const date = artifact.date ? ` on ${new Date(artifact.date).toLocaleDateString()}` : '';
    return `[${typeLabel}] ${title}${author}${date}`;
  }

  private classifyCoverage(entries: types.EvidencePack['entries']): 'sufficient' | 'partial' | 'insufficient' {
    const directCount = entries.filter(e => e.isDirect).length;
    const totalScore = entries.reduce((sum, e) => sum + e.relevanceScore, 0);

    if (directCount >= 3 && totalScore > 3.0) return 'sufficient';
    if (directCount >= 1 && totalScore > 1.0) return 'partial';
    return 'insufficient';
  }

  private identifyGaps(entries: types.EvidencePack['entries'], queryText: string): string[] {
    const gaps: string[] = [];
    const types = new Set(entries.map(e => e.artifact.artifact_type));

    if (!types.has('pr') && !types.has('issue')) {
      gaps.push('No direct PR or issue discussion found. Rationale may be inferred from commit messages.');
    }
    if (!types.has('adr')) {
      gaps.push('No Architecture Decision Records found. Explicit architectural rationale may be missing.');
    }
    if (entries.length < 3) {
      gaps.push('Limited evidence sources. Consider broadening the search.');
    }

    return gaps;
  }
}
